src/plus/integrations/providerIntegration.ts

```diff
@@ -26,22 +26,30 @@ import type {
 	PagedProjectInput,
 	PagedRepoInput,
 	ProviderAccount,
+	ProviderId,
 	ProviderIssue,
 	ProviderPullRequest,
 	ProviderRepoInput,
 	ProviderReposInput,
+	SelfHostedProviderId,
 } from './providers/models';
-import { IssueFilter, PagingMode, ProviderId, PullRequestFilter } from './providers/models';
+import { HostedProviderId, IssueFilter, PagingMode, PullRequestFilter } from './providers/models';
 import type { ProvidersApi } from './providers/providersApi';
 
-// TODO@eamodio revisit how once authenticated, all remotes are always connected, even after a restart
-
 export type SupportedProviderIds = ProviderId;
-export type ProviderKey = `${SupportedProviderIds}|${string}`;
+export type SupportedHostedProviderIds = HostedProviderId;
+export type SupportedSelfHostedProviderIds = SelfHostedProviderId;
+export type ProviderKey = `${SupportedHostedProviderIds}` | `${SupportedSelfHostedProviderIds}:${string}`;
+export type ProviderKeyById<T extends SupportedProviderIds> = T extends SupportedHostedProviderIds
+	? `${SupportedHostedProviderIds}`
+	: `${SupportedSelfHostedProviderIds}:${string}`;
 
 export type RepositoryDescriptor = { key: string } & Record<string, unknown>;
 
-export abstract class ProviderIntegration<T extends RepositoryDescriptor = RepositoryDescriptor> {
+export abstract class ProviderIntegration<
+	ID extends SupportedProviderIds = SupportedProviderIds,
+	T extends RepositoryDescriptor = RepositoryDescriptor,
```

Comment:
While we are refactoring all this, can we do away with requiring a repository descriptor for an integration? Certain integrations (like Jira, which I'm looking at now) don't support repositories - they're just issue providers that work with boards or projects.

Reply:
I'm not sure the implications of removing that yet (haven't had the time to look), but I would suggest doing that outside of this.

Reply:
Ok, no worries, I can include that in the Jira stuff.